src/accelerate/utils/deprecation.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import textwrap
+import warnings
+from typing import Callable, TypeVar
+
+from typing_extensions import ParamSpec
+
+
+_T = TypeVar("_T")
+_P = ParamSpec("_P")
+
+
+def deprecated(since: str, removed_in: str, instructions: str) -> Callable[[Callable[_P, _T]], Callable[_P, _T]]:
+    """Marks functions as deprecated.
+
+    It will result in a warning when the function is called and a note in the docstring.
+
+    Args:
+        since (`str`):
+            The version when the function was first deprecated.
+        removed_in (`str`):
+            The version when the function will be removed.
+        instructions (`str`):
+            The action users should take.
```

Comment:
Just a nit: I'd find it more intuitive if the `instructions` was renamed `instruction` and if the `"Please "` was not added in front, as I would assume as a caller that I need to pass a full sentence here.

Reply:
Done in [a7a43aa](https://github.com/huggingface/accelerate/pull/3161/commits/a7a43aa2f2268885b5db25a1324b04b59921ec65) and [2cb4648](https://github.com/huggingface/accelerate/pull/3161/commits/2cb46485f9bfda5c6108978fe786eeda3227ef0f) Thank you :)